### ISSUE TITLE
Add Option to set OpenBaseDir manually for domain/subdomain

### ISSUE THREAD

**Issue by dirkey** [Froxlor/Froxlor#515]
The OpenBaseDir can only be set to the "Home Directory" or the "Document Root". It would be nice to have a third option to set it manually. For example: 
Magento needs only it's pub directory to be accessible via Apache (DocumentRoot), but PHP needs access to all files in the "Mangento Root" (OpenBaseDir). Currently I must give it access to all files under the home directory of the user.

**Comment by d00p** [comment]
So, domain-documentroot is e.g. **/[user]/magento/pub** and your chice would be either **/[user]/** or **/[user]/magento/pub** but you need **/[user]/magento** - correct?

**Comment by d00p** [comment]
One of the "minor" problems here is that we try to avoid letting the customer enter paths manually as it might void the openbasedir functionality which the admin might wanted intentionally..that's why it's only these two choices

**Comment by dirkey** [comment]
I see! Maybe, we let users only allow to add relative paths to there home directory. But then, we need to take care of relative paths pointing to a parent directory. A solution is also to show the directory tree instead of a textbox. Then he is only allowed to choose these.

**Comment by NightmareJoker2** [comment]
My two cents: I'd love this. I've been creating separate users and using their home directories for it and that's become quite tedious to manage.
Allow the following three options: 
* document root
* parent of document root (must be inside home directory)
* home directory

If the parent of the document root is not inside the home directory, automatically set the home directory. (It's an absolute path string, so something like 
```
$home_path = realpath($home_directory);
$open_basedir_path = realpath($open_basedir);
if ($home_path === $open_basedir_path || substr($open_basedir_path, 0, strlen($home_path) + 1) === $home_path.DIRECTORY_SEPARATOR)
{
    // everything is fine
}
else
{
    $open_basedir_path = $home_path;
}
```
should suffice to make this safe.

This would help greatly with properly setting up Drupal, Magento, Tyop3 and custom built applications, such that only one small `index.php` and static files remain in the document root of the webs server's virtual host, and all application files, view templates (typically not with `php` extension, would be served as text), configuration, and possibly databases (e.g. SQLite ones) can safely remain outside of it and out of the reach of a FastCGI or rewrite engine failure of some sort from a package upgrade mishap.

**Comment by d00p** [comment]
Sounds good to me @NightmareJoker2

**Comment by florianschroen** [comment]
Hi, please review and merge my pull request https://github.com/Froxlor/Froxlor/pull/587, tested with v0.9.40.

**Comment by negrusti** [comment]
All Laravel projects need this too, and currently running with open_basedir disabled

**Comment by kakra** [comment]
Why don't we just decouple the `open_basedir` from the documentroot? I'd suggest the following solution:

Let the user choose between three settings per Domain:

  1. Use the customer home directory as `open_basedir` (and recommend against choosing this, it may have valid use-cases, tho)
  2. Use the domain root as `open_basedir` (which would default to `$HOME/$DOMAIN`, and should be the default choice)
  3. Use the document root as `open_basedir` (and that can only be located somewhere below the domain root)

This would need Froxlor to track another setting, i.e. what currently is `DocumentRoot` would become the domain root, and `DocumentRoot` would become a new setting.

Froxlor now just has to ensure that this new setting is:

  1. a relative path (thus not starting with `/`, or just hint that it will be appended as-is to `$domainroot`)
  2. ensure that `realpath($docroot)` starts with `realpath($domainroot)` (so that you can't escape `$domainroot` with symlinks)

**Comment by d00p** [comment]
chosing between customer-homedir and domain-docroot should be fair enough don't you think? that would be a simple select box. Starting with custom-values to be entered manually really complicates things

**Comment by d00p** [comment]
"expose" is a bit hard. Yes, a script can read/access the whole homedir of the customer. But require the customer to set a value there manually is imho not a good way. Depending what kind of customers you have but most of the "common webhosting customers" dont know what openbasedir is and don't care. 
Won't the solution suggested by @NightmareJoker2  in https://github.com/Froxlor/Froxlor/issues/515#issuecomment-410812127 be sufficient?

**Comment by negrusti** [comment]
By exposed I mean if one site is compromised, it will be trivial for
attacker  to compromise other sites on that account. I have a few Froxlor
servers where clients host several Laravel apps under one account.

**Comment by d00p** [comment]
I understand fully, and I can relate why this improvement is needed. Still, like said in https://github.com/Froxlor/Froxlor/issues/515#issuecomment-365264373 I try to avoid entering custom paths by customers as extensivly as possible

**Comment by Lantizia** [comment]
We were installing a copy of the very popular open-source learning platform [Moodle](https://moodle.org/) for a non-profit the other day.  The Moodle installation wizard wanted a path to store files which couldn't be accessed via the web.

This could be solved by having it outside of the DocumentRoot (I'll use Apache terminology in this comment), but that'd mean modifying the open_basedir to accommodate.  With Froxlor this unfortunately means an entirely new and separate PHP configuration would need creating & managing for just for this **one** single domain (as that extra path should **only** be for this specific copy of Moodle, a 2nd copy of Moodle on the same server would need its own!).  This is a pain when you want to tweak the base PHP config for all customer domains... you'll now have extra separate PHP configurations to keep in sync.

Unfortunately the way the Moodle installation wizard works at the moment... it won't proceed until that "Data directory" is outside of the DocumentRoot (even if it is properly secured with **.htaccess**).  I feel this is a bit crazy, so I logged a bug (Frolxor is mentioned a fair bit) about their docs being confusing and/or the security check being broken... [this bug here](https://tracker.moodle.org/browse/MDL-69337).  Ultimately I can see (on their forums and tracker) people on either side of the argument.

It strikes me... that we're likely to see more "off the shelf" content management platforms (e.g. Moodle, Magento, WordPress, Joomla, Drupal etc...) **or** extensions (installed within those sites)... that'll get more paranoid about where to store non-web accessible data.  Typically the only non-public sensitive data of much concern had been passwords (usually kept safely in a database)... but with more functionality (which traditionally needed a back end) being put into these CMS systems, they'll deal with more (and different kinds of) sensitive information.  They'll also likely get less concerned about offering up solutions for specific HTTP daemons (such as automatically creating Apache-specific **.htaccess** files for you) as the HTTP daemon space continues to diversify.

For this reason I'd love to see a way of **appending** to the **open_basedir** on a per-domain basis within the Froxlor panel.

I completely hear the point that @d00p made though about this shouldn't be a path that just anyone can arbitrarily state.

So how about for every DocumentRoot there is a same-named alternate private path...

Here are a few examples of how that might work...

| DocumentRoot  | Private path appended to open_basedir |
| ------------- | ------------- |
| /var/customers/webs/username/moodle.example.com/ | /var/customers/webs/private/username/moodle.example.com/  |
| /var/customers/webs/username/  | /var/customers/webs/username/private/  |
| /var/customers/webs/username/my-dir/something/moo/  | /var/customers/webs/username/private/my-dir/something/moo/  |

You'd have an option which can be enabled on a per-domain basis to enable / disable the "private path".  If you enable it and save... then it will automatically **mkdir -p** the same location used for DocumentRoot (minus the path for the username **/var/customers/webs/username/** part) within the 'private' subdirectory.

That way people can still get at the data via FTP/SFTP if required.

Thoughts?

**Comment by Lantizia** [comment]
Oh also (just out of interest) is there any reason why the setting "_Paths to append to OpenBasedir_" is under... **Settings | Webserver settings** and not in **Settings | PHP-FPM**, where one might expect it to be (since it's a **php.ini** directive)?

Or is it just because if you **weren't** using FPM (and instead used mod_php), you wouldn't actually have a **PHP**-named section under **Settings** to find PHP-related options?

**Comment by d00p** [comment]
Basically a good idea, but there will always be someone saying "but I need the folder to be named xyz" or whatever. 
So appending a folder within the customers home directory should be fair enough. I still think it's complicating things as other customers might need a folder within the domain docroot and so on and so forth. There's always another case that needs something else. Also, 90% of customers (end users!!) don't even know what open_basedir is...

So, if you know what you are doing, and not what you are installing, why not just deactivate open_basedir for the domain needing this?

**Comment by d00p** [comment]
> Oh also (just out of interest) is there any reason why the setting "_Paths to append to OpenBasedir_" is under... **Settings | Webserver settings** and not in **Settings | PHP-FPM**, where one might expect it to be (since it's a **php.ini** directive)?
> 
> Or is it just because if you **weren't** using FPM (and instead used mod_php), you wouldn't actually have a **PHP**-named section under **Settings** to find PHP-related options?
> 
> 

1) because it was there way before fcgid/fpm and never switched section but:
2) also yes, because you would not have the php-related setting-sections if you'd use mod_php

**Comment by negrusti** [comment]
I don't think it's a good idea to do it this way. Many sites are deployed from git, and document root is a subdirectory of the whole project.  Can't be restructured to fit this un-natural Froxlor quirk.

**Comment by d00p** [comment]
@negrusti you're welcome to provide a better solution and a pull-request :)

**Comment by Lantizia** [comment]
> So appending a folder within the customers home directory should be fair enough.

Yeah that'd be much simpler :)

Since we use this setting...

![image](https://user-images.githubusercontent.com/10448369/88708860-36f41500-d10c-11ea-86a3-3ab63f543d6c.png)

The DocumentRoot is automatically set (if it isn't overridden) as a sub-directory in their "home directory" with the same name as the websites domain.

So for us... something equally as simple... like a **_private** suffix on the same formula would work quite well...

**DocumentRoot:** /var/customers/webs/username/moodle.example.com
**Private per-website "Data directory" area:** /var/customers/webs/username/moodle.example.com_private

Also since many of these "installation wizards" that come with common PHP-written CMS's can **detect and then display** the absolute path they're running in (the DocumentRoot)... which **if left untouched** looks like this... 

![image](https://user-images.githubusercontent.com/10448369/88709716-74a56d80-d10d-11ea-8454-f606334da9fc.png)

It should then be really easy to explain to the user that they can just set that "Data directory" (or similar such name) to the same as the path their website detected it's running from (the DocumentRoot) if it's displayed.  They'll just need to add a **_private** (or whatever suffix is nominated in Froxlor) and it'll work without them getting any **open_basedir** errors (they wouldn't need to know what **open_basedir** even is!).

Visually explaining it would be as easy as...

![image](https://user-images.githubusercontent.com/10448369/88710068-fb5a4a80-d10d-11ea-8a5b-be4735c2baf0.png)

I can see this working for many situations, not just Moodle.  I just happen to have Moodle to hand for screenshot purposes... perhaps @dirkey knows what this looks like on a Magento installation?

The reason I'm saying all this, is because I'm dreaming up a new setting (to go alongside the first one I mentioned... in **Settings | System settings**) to make this even easier for everyone...

![image](https://user-images.githubusercontent.com/10448369/88711269-d961c780-d10f-11ea-90d1-822eb5ef7576.png)

Also, I like fake screenshots :)

**Comment by d00p** [comment]
I think that would be the best solution for both sides, nothing manually entered but works. Just has to be communicated with the customer (e.g. good setting description)

PS: yeah fake screenshots are great ;p

**Comment by d00p** [comment]
But remember, if it's in system-settings the customer cannot change it. It should be a customer option to enable a domain-based dara-dir (or whatever you wanna call it)

**Comment by Lantizia** [comment]
Yeah I agree, on the domain settings page... one checkbox to enable the "private area"... which if that is enabled... then lets you set a string to either 
**a)** set whatever dir to use for it (from the perspective of the users "home" dir) or...
**b)** override what that dir v.s. whatever **might** be set in the system-settings section

**Comment by Lantizia** [comment]
Yeah, so on the domain settings you'd have...

1) A checkbox to enable the "private area".  A nice friendly name for this function... but the actual checkbox's description can state that its actually to append to open_basedir a single path... creating it if non-existant and with a predetermined name if not overridden with...
2) A text box for the user to **optionally** enter a directory name (relative to the users "home" dir) to use with this function... if they enter something, then it overrides using domain name + suffix (suffix being set in the admin settings).  Obviously the string in this text box can be ignore if they didn't check the checkbox.

That way you could leave a default suffix like '_private' in system-settings without it affecting anyone.

**Comment by d00p** [comment]
then you would have customer input again which totally makes the checkbox unnecessary if you could just enter a path which still is something i'm trying to avoid (from a customers point of view). Thought we agreed on the checkbox so nothing needs to be entered manually, see https://github.com/Froxlor/Froxlor/issues/515#issuecomment-665282647

**Comment by Lantizia** [comment]
I'm happy with just a checkbox (and then it'd **always** be a extra path comprised of "domain + suffix")... but I was trying to accommodate anyone else who might want to specifically state which path (relative to the "home" dir) they want if they're not wanting that.

In way it's no different to how this setting works...

![image](https://user-images.githubusercontent.com/10448369/88798697-39eb1600-d19d-11ea-8189-d598118055cd.png)

As setting that, doesn't mean you can't override the DocumentRoot on a per-domain basis anyway... it just presumes an answer if that text input box isn't filled in.

**Comment by d00p** [comment]
keep in mind that many things you are refering to are admin-settings - customer cannot change or set these! there's a big difference what an admin can and should do and what a customer is allowed to do and should be able to

**Comment by Lantizia** [comment]
OK so looking at this from the perspective of a logged in customer... 

In these fake screenshots I've used non-technical language to describe the settings and tried to keep any terminology similar to language already used on that page...

![image](https://user-images.githubusercontent.com/10448369/88802325-d19f3300-d1a2-11ea-9401-9d403cb3cfd0.png)

Which since (in this particular case) our **Path:** (not Private path) is the name of the domain (which was automatically set), then leaving the private area path completely blank... would do **exactly** the same job...

![image](https://user-images.githubusercontent.com/10448369/88801564-cdbee100-d1a1-11ea-9603-0f93e32022ab.png)

But if someone doesn't like the "domain + suffix" path, they can override it...

![image](https://user-images.githubusercontent.com/10448369/88801689-f1822700-d1a1-11ea-8258-f7002ed09dbe.png)

If someone doesn't want to append an extra path to **open_basedir** at all... they just want the normal **open_basedir** which is to include the **Path:** (i.e. the DocumentRoot)... then they can just turn this off (which should be the default)...

![image](https://user-images.githubusercontent.com/10448369/88802582-2c388f00-d1a3-11ea-9e89-dd43a7c5a7df.png)

However if someone wants the "private area" to actually just be the entirety of the users "home" directory... (setting '/' is different to blank)

![image](https://user-images.githubusercontent.com/10448369/88802698-59853d00-d1a3-11ea-860c-49c03a5a246c.png)

Which incidentally means that the following section... which is already in Froxlor (it's just been missing from all the screenshots above)... can now be **completely** removed... 

![image](https://user-images.githubusercontent.com/10448369/88803042-cdbfe080-d1a3-11ea-8543-e157b10e097f.png)

As we already have ways of achieving both options which it had in it's drop down box.  It's also better explained and is less "scary" sounding than the way it is presented now.

**Comment by d00p** [comment]
> Which incidentally means that the following section... which is already in Froxlor (it's just been missing from all the screenshots above)... can now be completely removed...

no? Private path is **appended** it does not replace domain-docroot or customer-homedir as openbasedir-path

**Comment by Lantizia** [comment]
Right but I'm saying whether private path is appended or not... the open_basedir will always contain the documentroot.

**Comment by d00p** [comment]
no, it will not. If openbasedir-path is set to home-directory and no private area is activated it will only contain the customers-homedir ...not the domains documentroot.

Anyway, I think we're on a good way, still thinking of something to avoid the need for manual (customer) action. I liked the idea to "standarize" the private-data directory using domain-name and a suffix - just unsure whether all systems (moodle can, others might not) are able to use a customized data-directory

**Comment by Lantizia** [comment]
I'm saying **change** the default behaviour... that if no private area is activated... the **open_basedir** should contain only the domain-documentroot... and not the customers-homedir.

**Comment by Lantizia** [comment]
Actually I'm not saying change it at all... that already is the default on my system.

**Comment by Lantizia** [comment]
> just unsure whether all systems (moodle can, others might not) are able to use a customized data-directory

Well it wouldn't matter if they didn't support a customer data-directory if the panel lets the customer choose it also.

**Comment by d00p** [comment]
So why not name it like "directories to append to open-basedir" and allow customers to specify whatever directory (whithin their homedir) they want? That's somehow exactly that...

**Comment by d00p** [comment]
> One of the "minor" problems here is that we try to avoid letting the customer enter paths manually as it might void the openbasedir functionality which the admin might wanted intentionally..that's why it's only these two choices

my initial answer still is valid though, an admin might not want that so you would again have additional options (on a per customer base, or even domain based) whether or not to allow adding custom openbasedir-values

**Comment by Lantizia** [comment]
Because we were trying to strike a balance between automatically picking a sensible directory for the customer without manual entry??... and also trying not to use terminology they may not full understand (e.g. moodle's wizard never said "Directory for Moodle data (within your open_basedir paths if using open_basedir):"... it just says... "Data directory:" :P)

**Comment by Lantizia** [comment]
> my initial answer still is valid though, an admin might not want that so you would again have additional options (on a per customer base, or even domain based) whether or not to allow adding custom openbasedir-values

But these paths are **relative** to their own home directory!!  What admin would care if they add one single open_basedir path per domain - when it's a path inside of their own user space?!

**Comment by d00p** [comment]
these paths are not relative to the homedir directory, one is the other one **is** the home-directory.

Well then, you seem that you have given that deep thought, so please, provide a well tested pull-request and I gladly merge it.

**Comment by Lantizia** [comment]
> these paths are not relative to the homedir directory, one is the other one **is** the home-directory.

They way they'd be entered into this form **they would be** (see the descriptions of the settings in the faked screenshots)... that is part of the whole point.

**Comment by d00p** [comment]
Yeah yeah, got you. You were talking about your "private path field".
And yes, most admins wont care - other might. I'm just saying what I think out of experience with new options from the last 10 years.

As said, I'll gladly merge any pull-request provided if tested thoroughly

**Comment by negrusti** [comment]
I think the solution that will cover 99% of use cases is 3 options for open_basedir: 
1) document_root
2) 1 level above document_root
3) off

**Comment by d00p** [comment]
"Off" ist no option there for a customer. That's an admin setting (openbasedir on/off). The openbasedir path which can be selected is either homedir or domain-documentroot.

A third option with "one level above domain documentroot" could work yes

**Comment by Lantizia** [comment]
>     2\. 1 level above document_root

I guess if you picked that option... and had something like this as document_root...

**/domain.tld/htdocs/**     (or httpdocs, www,  etc...)

Means you could then choose to have **something like** this as where private data can go (for use by the website running in the document_root only)...

**/domain.tld/private/**

Four problems I see though (at least)...

1. That new setting doesn't work if the document_root itself is '/'... as in... the "home" directory itself... otherwise you're allowing customers to access other customers data.
2. That private directory won't be made created, so it's more work explaining to the customer that they need to create it via SFTP/FTP first before they can then give the path to their CMS's installation wizard.
3. You'd ideally want a new optional global setting... which doesn't just auto-populate the default "Path:" field (document_root) as the domain name (like it can now)... but also adds this extra sub-directory... or you'll be forever moving peoples website data around to accommodate the need for a private directory - when the customer ultimately realises they need one later on (especially if they install some crazy plugin/extension down the line that wants one).
4. Any existing installations of Froxlor which already uses the domain name in the "Path:" field (document_root)... would need to move the website contents around on the filesystem to take advantage of this... where as my proposal above wouldn't as it's a directory that sits alongside.

But yeah as @d00p said, you wouldn't want an 'Off' option as that'd not be very secure letting them decide that.  So this would be a third option to add to the "Home dir" or "Document root" options.  But I personally feel the "Document root" should be in open_basedir regardless of any option chosen... as when would you **not** want that!?

**Comment by Lantizia** [comment]
Oh also this is starting to sound a little bit like Plesk (sorry it's the control panel that I'm exposed to the most in my day job for an ISP)... where the default layout is like this...

```
user@box:/var/www/vhosts/domain.tld# ls -lah
total 20K
drwx--x---  5 domain psaserv 4.0K Feb 25  2019 .
drwxr-xr-x 14 root root    4.0K Jun  5  2019 ..
drwxr-xr-x  2 domain psacln  4.0K Aug 17  2017 error_docs
drwxr-xr-x 15 domain psacln  4.0K Nov 14  2017 httpdocs
drwx------  2 domain root    4.0K Jul 29 06:25 logs
```

**Comment by negrusti** [comment]
@Lantizia point 1 is incorrect - by default 1 level above is a customer home directory, no other customer dirs can be accessed (and they have different ownership anyway).
Re. 4) you are trying to fit this solution to Moodle setup wizard. Most of the projects requiring this open_basedir fix will have a fixed structure as opposed to an arbitrary Moodle data dir. Let's not implement this in a Moodle-specific way.

**Comment by NightmareJoker2** [comment]
@Lantizia Since this is basically going to turn back into what I wrote at https://github.com/Froxlor/Froxlor/issues/515#issuecomment-410812127, I'll offer some more feedback:

> 1. That new setting doesn't work if the document_root itself is '/'... as in... the "home" directory itself... otherwise you're allowing customers to access other customers data.

No, it won't. You just have to make sure a customer's document root and `open_basedir` path is always inside their home directory or set it to the home directory otherwise and be done with it.

> 2. That private directory won't be automatically created, so it's more work explaining to the customer that they might need to create it via SFTP/FTP first before they can then give the path to their CMS's installation wizard (assuming that won't create it).

That's not necessary. You should know what you are installing and where, and if you need a specific folder structure, be perfectly capable of creating it yourself. It makes no sense to create these folders by default, because you can't be sure you'll need them or that they're even going in the right place.

> 3. You'd ideally want a new optional global setting... which doesn't just auto-populate the default "Path:" field (document_root) as the domain name (like it can now)... but also adds this extra sub-directory (htdocs, httpdocs, www, or whatever)... or you'll be forever moving peoples website data around to accommodate the need for a private directory - when the customer ultimately realises they need one later on (especially if they install some crazy plugin/extension down the line that wants one).  Which means ultimately another global setting... for making this new "1 level above document_root" option - the default one :)

No, you don't. Making a default directory with the domain name, setting it as the document root for the domain and adding an index document makes sense, because as soon as the web server configuration is updates, there will be something there to tell you that the newly configured domain works. Creating additional structures just confuses users who don't need them. Making the *one level above document root* option the new default is potentially dangerous. You can't know in advance that there's nothing already there that might end up exposed. The default should always be the option that is the most restrictive.

> 4. Any existing installations of Froxlor which already uses the domain name in the "Path:" field (document_root)... would need to move the website contents around on the filesystem to take advantage of this... where as my proposal above wouldn't as it's a directory that sits alongside.

But your method doesn't work for everyone, and not at all for any of the popular things.

> 5. I feel it still requires the customer needing to know **more** about **open_basedir** (and would require a good explaination of the consequences of each of the 3 options) than simply humanising it bit with the idea of a **private area** that is suggested for them

How about adding a link to https://www.php.net/manual/en/ini.core.php#ini.open-basedir (and their other language equivalents in the respective translations) next to the setting? Though, I would think people can use Google...

In any case, since the feature doesn't yet exist, and you sort of need it, just create a separate customer account for every website that needs a directory outside of the document root to function, set the `open_basedir` path to *Home Directory* and create your "private directory" wherever you like. This works well, is update-safe, fully supported, and you don't need to mess with any of the global configuration or site specific PHP configurations, either.

**Comment by Lantizia** [comment]
@negrusti said...
> point 1 is incorrect - by default 1 level above is a customer home directory, no other customer dirs can be accessed (and they have different ownership anyway)

This isn't right... at least on our system the **/var/customers/webs** dir is drwxr-xr-x... so they can **at least** list all other customers usernames.  Additionally if we hadn't chosen to map Froxlor users to system users (e.g. no need for PHP-FPM) then who knows what these would be owned as and the consequences of that?  But as @NightmareJoker2 said in his **first** comment on this issue (6th August 2018), you'd just add a security check to ensure it's inside the home directory first... and this issue goes away.

@NightmareJoker2 said...
> How about adding a link to https://www.php.net/manual/en/ini.core.php#ini.open-basedir (and their other language equivalents in the respective translations) next to the setting? Though, I would think people can use Google...

I might be wrong, but I think this exactly what @d00p would prefer to avoid... giving the customers more stuff to learn and a more complicated experience.  On the other hand, we've already got an option (that the customers see when editing a domain) called **OpenBasedir-path:** with absolutely no explanation of what that really means!

@NightmareJoker2 said...
> since the feature doesn't yet exist, and you sort of need it

Wrong, I don't.  I started to add my thoughts to this issue **in case** the guys over at Moodle didn't come around to the view that **.htaccess** is secure enough ([in a bug I logged with them](https://tracker.moodle.org/browse/MDL-69337)).  But I see the possibility of other CMS-like systems being this paranoid, either now or in the future (and remember this started with @dirkey mentioning Magento).  I'm just trying to future proof Froxlor... I don't actually need this feature right now.

@NightmareJoker2 said...
> just create a separate customer account for every website

I actually love this idea for more than just the reason you suggested it... I even brought this up with @d00p on IRC months and months ago.  I like how every website would get it's own system user which adds to security (further preventing the issue of websites owned by the same customer from meddling with each other).  I've even renamed 'customers' to 'subscriptions' in my Froxlor UI (mostly to see how it'd look, as a test) and then I think the idea was **along the lines of** (this will miss a lot of detail as it's not the point of this issue we're discussing) to give out admin-type logins to customers, but restricted to just their particular subscriptions (renamed from 'customers').  Ultimately I was told this feature proposal, would make it too much like Plesk... which whilst Plesk does suck in many ways, I still don't think this is a bad idea :)

@negrusti said...
> Can't be restructured to fit this un-natural Froxlor quirk.

@negrusti also said...
> Re. 4) you are trying to fit this solution to Moodle setup wizard. Most of the projects requiring this open_basedir fix will have a fixed structure as opposed to an arbitrary Moodle data dir. Let's not implement this in a Moodle-specific way.

Twice now you've implied I'm intending to alter things with a narrow view point, that's a bit of a leap!  If I only cared about Moodle (which as you point out, lets you pick a "data directory" path) then I wouldn't care about adding a text field (much to @d00p's disappointment, who I think would prefer to do without it) for entering a custom (but relative to home) path, now would I?  If anything I'm trying to accommodate **all of the requirements** we've seen on this issue so far... even if it means re-hashing some old ones from over 2 years ago (e.g. @dirkey was actually the first to suggest the paths should be relative only).

@NightmareJoker2 said...
> your method doesn't work for everyone, and not at all for any of the popular things

You might be completely right.  But can you give some examples (perhaps CMS's that people often use?) where it would be difficult for the customer to set a "data directory" that sits **alongside** the document_root rather than one level below it?

@NightmareJoker2 said...
> You should know what you are installing and where

I should sure, but we're not on about me... this is loosing sight of the issue!  If it was just me I wouldn't need Froxlor at all!

@NightmareJoker2 said...
> Creating additional structures just confuses users who don't need them

Actually these subdirectories are quite normal with all shared hosting utilising the worlds most popular 2 control panels... where Plesk uses httpdocs and cPanel uses public_html.  I'm not saying this **reason alone** is good enough to have them... but then I also wasn't actually mandating we do it, it was a suggestion for two **OPTIONAL** settings.  One to use "domain + subdir" as document_root (**subdir** being www / pub / httpdoss / htdocs / public_html / whatever) rather than the current (also optional) setting to just use "domain" as document_root.  And a second **optional** setting to make the new "one level above document_root as open_basedir" the default, so people can take advantage of it straight away without needing **to do anything**.

**In conclusion**...

I'm sure the discussion will continue... but I'm personally not fussed which way we do it as long as it is done properly, is thoroughly thought out and easy to explain to the customer.

**Comment by negrusti** [comment]
@Lantizia default document root is
/var/customers/webs/<username>/<domain name>
1 level above is
/var/customers/webs/<username>

**Comment by d00p** [comment]
wrong, default is `/var/customers/webs/[user]` - the domainname is only included if you have enabled the corresponding setting to append it to the homedir

**Comment by negrusti** [comment]
@d00p ok my bad then - but how is it a sensible default?

**Comment by d00p** [comment]
how is it not? You have you home-directory. End of story. If an admin enabled openbasedir, you are restricted to it. How come this is not a sensible default?

In terms of apps that require a data directory outside the public-htdocs - yes you are right. But that's a usage thing

**Comment by negrusti** [comment]
@d00p customers can, and often do have multiple domains - I don't see the point of default document root not being named as the domain name.

**Comment by d00p** [comment]
well then YOU as admin should enable the setting to create subdirectories for every domain. that's why it is a setting. Different behaviour for different needs.

**Comment by Lantizia** [comment]
Had a long chat about this on IRC with @d00p, just pasting one of the main bits of agreement here at his request, just so we don't loose it...

Assuming the administrative setting "Use domain name as default value for DocumentRoot path" turns into a drop down box (instead of a check box) and renamed to something like "Default DocumentRoot path" then you could have these 4 options... 

default **doc_root path** possible options...       **0** => "/"  **1** => "/domain.tld/"  **2**=> "/domain.tld/htdocs/"  **3**=> "/htdocs/"

Notes...
- Not suggesting we call the drop down options **exactly** as shown, it's just for illustrative purposes.
- Compared to the existing checkbox we already have... 0 is the equivalent of unticked and 1 is the equivalent of ticked
- We can add an extra setting (a text field) for specifying what 'htdocs' should actually be so people can use "public_html" / "www" / "pub" / "httpdocs" / whatever...

Assuming we also make a new administrative setting of "Default OpenBasedir-path policy" which would be a drop down box as well, with these 3 options...

default **openbase_dir policy**... **0** => "doc_root itself"  **1** => "home dir"  **2** => "parent of doc_root"

Then the following scenarios (dependant on how the administrator prefers to run their web server) are all possible (but not limited to, just the ones I can think of - off the top of my head)... 

Also the expression "private areas" in the below context... basically means any path outside of the DocumentRoot to stick files within their home directory.

- If the admin allows only 1 domain per customer & presumes no one wants private areas... sets these defaults... **doc_root path** = 0 and **openbase_dir policy** = 0
- If the admin allows only 1 domain per customer & wants it easier to use private areas... sets these defaults... **doc_root path** = 3 and **openbase_dir policy** = 1
- If the admin allows more than 1 domain per customer, presumes customers will set their own doc_root & presumes no one wants private areas... sets these defaults... **doc_root path** = 0 and **openbase_dir policy** = 0
- If the admin allows more than 1 domain per customer, presumes customers will set their own doc_root & wants it easier to use private areas... sets these defaults... **doc_root path** = 3 and **openbase_dir policy** = 2
- If the admin allows more than 1 domain per customer & presumes no one wants private areas... sets these defaults... **doc_root path** = 1 and **openbase_dir policy** = 0
- If the admin allows more than 1 domain per customer & wants it easier to use private areas... sets these defaults... **doc_root path** = 2 and **openbase_dir policy** = 2

So basically the 4 options of **doc_root path** and 3 options of **openbase_dir policy** all have a purpose and make more sense when appropriately combined together

**Comment by smaddy** [comment]
anyone found a quick workaround?

**Comment by d00p** [comment]
a "quick" workaround is disable open_basedir for the domain
